### PR TITLE
Tiered index add vector MOD-4316

### PR DESCRIPTION
**Pull request by alonre24**
Implement `addVector` for the tiered index, where we **assume that there are no overwrites** (for now). Insertion is done into the flat buffer, the insert job callback that is responsible for adding the vector into HNSW graph is not part of this PR scope.
This includes some refactoring of the tiered index structures - among them, making the `AsyncJob` and its derivatives jobs structs inherit from `VecSimObjectBase`, so we can account for the memory that is allocated for the jobs

### CODE REVIEW

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 1]
```diff
@@ -24,34 +61,32 @@ class TieredHNSWIndex : public VecSimTieredIndex<DataType, DistType> {
...
     // Wrappers static functions to be sent as callbacks upon creating the jobs (since members
     // functions cannot serve as callback, this serve as the "gateway" to the appropriate index).
-    static void executeInsertJobWrapper(HNSWInsertJob *job) {
-        reinterpret_cast<TieredHNSWIndex<DataType, DistType> *>(job->index)->executeInsertJob(job);
-    }
-    static void executeRepairJobWrapper(HNSWRepairJob *job) {
-        reinterpret_cast<TieredHNSWIndex<DataType, DistType> *>(job->index)->executeRepairJob(job);
-    }
+    static void executeInsertJobWrapper(void *job);
+    static void executeRepairJobWrapper(void *job) {}
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_tiered.h:65]: temporary right?

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 2]
```diff
@@ -5,15 +5,52 @@
...
+                  JobCallback insertCb, void *index_)
+        : AsyncJob(allocator, HNSW_INSERT_VECTOR_JOB, insertCb), index(index_), label(label_),
+          id(id_) {}
+};
+
+/**
+ * Definition of a job that swaps last id with a deleted id in HNSW Index after delete operation.
+ */
+struct HNSWSwapJob : public AsyncJob {
+    void *index;
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_tiered.h:26]: Why not `VecSimIndex`?

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 3]
```diff
@@ -5,15 +5,52 @@
...
+                                      // is ready to be executed (atomic counter).
+    // TODO: implement contractor
+};
+
+/**
+ * Definition of a job that repairs a certain node's connection in HNSW Index after delete
+ * operation.
+ */
+struct HNSWRepairJob : public AsyncJob {
+    void *index;
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_tiered.h:38]: Why not `VecSimIndex`?

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 4]
```diff
@@ -5,15 +5,52 @@
 #include "hnsw_factory.h"
 
 #include <unordered_map>
+/**
+ * Definition of a job that inserts a new vector from flat into HNSW Index.
+ */
+struct HNSWInsertJob : public AsyncJob {
+    void *index;
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_tiered.h:12]: Why not `VecSimIndex`?

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 5]
```diff
@@ -5,15 +5,52 @@
...
+
+/**
+ * Definition of a job that repairs a certain node's connection in HNSW Index after delete
+ * operation.
+ */
+struct HNSWRepairJob : public AsyncJob {
+    void *index;
+    idType node_id;
+    unsigned short level;
+    HNSWSwapJob *associated_swap_job;
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_tiered.h:41]: Didn't we say that a repair job should be associated with more than one swap job?

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 6]
```diff
@@ -73,3 +108,90 @@ class TieredHNSWIndex : public VecSimTieredIndex<DataType, DistType> {
...
+}
+
+template <typename DataType, typename DistType>
+int TieredHNSWIndex<DataType, DistType>::addVector(const void *blob, labelType label,
+                                                   bool overwrite_allowed) {
+    /* Note: this currently doesn't support overriding (assuming that the label doesn't exist)! */
+    this->flatIndexGuard.lock();
+    if (this->flatBuffer->indexCapacity() == this->flatBuffer->indexSize()) {
+        this->flatBuffer->increaseCapacity();
+    }
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_tiered.h:175]: Why is this needed? isn't that part of `addVector`?

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 7]
```diff
@@ -73,3 +108,90 @@ class TieredHNSWIndex : public VecSimTieredIndex<DataType, DistType> {
...
+    AsyncJob *new_insert_job = new (this->allocator)
+        HNSWInsertJob(this->allocator, label, new_id, executeInsertJobWrapper, this);
+    // Save a pointer to the job, so that if the vector is overwritten, we'll have an indication.
+    if (this->labelToInsertJobs.find(label) != this->labelToInsertJobs.end()) {
+        // There's already a pending insert job for this label, add another one (without overwrite,
+        // only possible in multi index)
+        assert(this->index->isMultiValue());
+        this->labelToInsertJobs.at(label).push_back((HNSWInsertJob *)new_insert_job);
+    } else {
+        auto new_jobs_vec = vecsim_stl::vector<HNSWInsertJob *>(1, (HNSWInsertJob *)new_insert_job,
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_tiered.h:187]: this is using a copy constructor, right?

File: tests/unit/test_hnsw_tiered.cpp [hunk 8]
```diff
@@ -78,11 +74,94 @@ TYPED_TEST(HNSWTieredIndexTest, CreateIndexInstance) {
...
-        ASSERT_EQ(tiered_index->labelToInsertJobs[vector_label].size(), 0);
+        ASSERT_EQ(tiered_index->labelToInsertJobs.at(vector_label).size(), 0);
+
+        // Cleanup.
+        delete jobQ;
+        VecSimIndex_Free(tiered_index);
+    }
+}
+
+TYPED_TEST(HNSWTieredIndexTest, addVector) {
```
Comment (GuyAv46) [inline tests/unit/test_hnsw_tiered.cpp:85]: Nice test! but don't you want to validate the index state after performing the insertion job?

**Review by GuyAv46 — COMMENTED** [review]
Nice! Few questions

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h (hunk 5 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:41]: No, it's the other way around (multiple repair jobs are associated with a single swap job)

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h (hunk 6 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:175]: No, we intentionally moved the `resize` outside the `addVector`, as the `resize` requires a write lock in HNSW, while add vector is done under read lock. Hence, the generic API now has a separate `increaseCapacity` call (which is used in flat index as well)

File: tests/unit/test_hnsw_tiered.cpp (hunk 8 above)
Comment (alonre24) [reply tests/unit/test_hnsw_tiered.cpp:85]: Implementing the insertion background job is not part of this PR (but it's coming... very soon...)

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h (hunk 7 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:187]: Yep. I guess I can avoid the copy and just create the vector instead.

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h (hunk 2 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:26]: I guess we can

**Review by GuyAv46 — APPROVED** [review]

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 9]
```diff
@@ -5,15 +5,52 @@
...
+    HNSWInsertJob(std::shared_ptr<VecSimAllocator> allocator, labelType label_, idType id_,
+                  JobCallback insertCb, VecSimIndex *index_)
+        : AsyncJob(allocator, HNSW_INSERT_VECTOR_JOB, insertCb), index(index_), label(label_),
+          id(id_) {}
+};
+
+/**
+ * Definition of a job that swaps last id with a deleted id in HNSW Index after delete operation.
+ */
+struct HNSWSwapJob : public AsyncJob {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:25]: right

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 10]
```diff
@@ -5,15 +5,52 @@
 #include "hnsw_factory.h"
 
 #include <unordered_map>
+/**
+ * Definition of a job that inserts a new vector from flat into HNSW Index.
+ */
+struct HNSWInsertJob : public AsyncJob {
+    VecSimIndex *index;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:12]: Why so? There is no shared ownership... It's more like a week reference IMO... So that is the index deleted, we just drop the job.

File: src/VecSim/vec_sim_tiered_index.h [hunk 11]
```diff
@@ -5,6 +5,17 @@
...
+/**
+ * Definition of generic job structure for asynchronous tiered index.
+ */
+struct AsyncJob : public VecsimBaseObject {
+    JobType jobType;
+    JobCallback Execute; // A callback that receives a job as its input and executes the job.
+
+    AsyncJob(std::shared_ptr<VecSimAllocator> allocator, JobType type, JobCallback callback)
+        : VecsimBaseObject(allocator), jobType(type), Execute(callback) {}
+};
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/vec_sim_tiered_index.h:17]: What do you mean? It is a general struct no?

**Review by DvirDukhan — APPROVED** [review]
